torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -164,16 +176,26 @@ at::Tensor DoBinaryOp(const at::Tensor& self, const at::Scalar& other,
   return bridge::AtenFromXlaTensor(result);
 }
 
-void CheckBinaryOpTypePromotion(const at::Tensor& out, const at::Tensor& self,
-                                const at::Tensor& other) {
-  at::ScalarType resultType = at::result_type(self, other);
-  XLA_CHECK(at::canCast(/*from=*/resultType, /*to=*/out.scalar_type()));
+template <typename B>
+at::Tensor DoBinaryOpWithoutPromo(const at::Tensor& self,
+                                  const at::Tensor& other, const B& bin_op) {
+  at::ScalarType dtype = at::result_type(self, other);
+  XLA_CHECK(at::canCast(/*from=*/dtype, /*to=*/self.scalar_type()));
```

Comment:
I'm not quite certain how this works. Why do we need to compute if the promoted type can cast to the scalar type of self? (At the very least, this is asymmetric; what about other? But it seems to me like this ought never to fail)

Reply:
I think this is meant to mirror the in-place/out= check that TensorIterator has here: https://github.com/pytorch/pytorch/blob/05c1e5b65571d2fa14b0f06c37ba970fd7cc43d0/aten/src/ATen/TensorIterator.cpp#L405.

It does seem overly restrictive to put that check here though, since this code gets called by all variants (functional, inplace, out=), and it shouldn't fail in the functional case. E.g.
```
>>> a = torch.ones(2, dtype=torch.float32)
>>> b = torch.ones(2, dtype=torch.complex128)
>>> a.add(b)
tensor([3.+0.j, 3.+0.j], dtype=torch.complex128)
>>> a.add_(b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: result type ComplexDouble can't be cast to the desired output type Float
```

This seems like something that https://github.com/pytorch/pytorch/issues/58765 should solve? And for now we should just remove this check?

An alternative would be to codegen `meta` calls into the pt/xla, like we do for in-tree, but that would only solve the problem for TensorIterator kernels (and also calculate output sizes too, which would be wasted work - and I remember from a few weeks ago that we didn't really want to go down this route).

Reply:
OK, satisfied by this explanation.

Reply:
Agree, I will remove this check.